packages/valory/skills/abstract_round_abci/tests/test_tools/test_rounds.py

```diff
@@ -0,0 +1,689 @@
+        base_round_test = BaseRoundTestClass()
+        base_round_test._event_class = DummyEvent  # pylint: disable=protected-access
+
+        base_round_test._test_no_majority_event(  # pylint: disable=protected-access
+            MagicMock(
+                end_block=lambda: (
+                    MagicMock(),
+                    DummyEvent.NO_MAJORITY,
+                )
+            )
+        )
+
+    @staticmethod
+    @given(st.integers(min_value=0, max_value=100), st.integers(min_value=1))
+    def test_complete_run(iter_count: int, shift: int) -> None:
+        """Test `_complete_run`."""
+
+        def dummy_gen() -> Generator[MagicMock, None, None]:
+            """A dummy generator."""
+            return (MagicMock() for _ in range(iter_count))
+
+        # test with the same number as the generator's contents
+        gen = dummy_gen()
+        BaseRoundTestClass._complete_run(  # pylint: disable=protected-access
+            gen, iter_count
+        )
+
+        # assert that the generator has been fully consumed
+        with pytest.raises(StopIteration):
+            next(gen)
+
+        # test with a larger count than a generator's
+        with pytest.raises(StopIteration):
+            BaseRoundTestClass._complete_run(  # pylint: disable=protected-access
+                dummy_gen(), iter_count + shift
+            )
+
+
+class BaseTestBase:  # pylint: disable=too-few-public-methods
+    """Base class for the Base tests."""
+
+    base_round_test: BaseRoundTestClass
+    base_round_test_cls: Type[BaseRoundTestClass]
+
+    def setup(self) -> None:
+        """Setup that is run before each test."""
+        self.base_round_test = self.base_round_test_cls()
+        self.base_round_test._synchronized_data_class = (  # pylint: disable=protected-access
+            DummySynchronizedData
+        )
+        self.base_round_test.setup()
+        self.base_round_test._event_class = (  # pylint: disable=protected-access
+            DummyEvent
+        )
+        self.base_round_test.consensus_params._max_participants = (  # pylint: disable=protected-access
+            MAX_PARTICIPANTS
+        )
+
+
+class DummyCollectDifferentUntilAllRoundWithEndBlock(
+    DummyCollectDifferentUntilAllRound
+):
+    """A `DummyCollectDifferentUntilAllRound` with `end_block` implemented."""
+
+    def __init__(self, dummy_exit_event: DummyEvent, *args: Any, **kwargs: Any):
+        """Initialize the dummy class."""
+        super().__init__(*args, **kwargs)
+        self.dummy_exit_event = dummy_exit_event
+
+    def end_block(self) -> Optional[Tuple[BaseSynchronizedData, Enum]]:
+        """A dummy `end_block` implementation."""
+        if self.collection_threshold_reached and self.dummy_exit_event is not None:
+            return (
+                cast(
+                    DummySynchronizedData,
+                    self.synchronized_data.update(
+                        most_voted_keeper_address=list(self.collection.keys())
```

Comment:
probably doesn't matter, but this should be the value instead of the key.
`most_voted_keeper_address=set(self.collection.values()).pop()`